executor/src/witgen/jit/block_machine_processor.rs

```diff
@@ -125,22 +131,123 @@ impl<'a, T: FieldElement> BlockMachineProcessor<'a, T> {
                     .iter()
-                    .filter(|id| is_machine_call(id))
-                    .map(move |id| (id, row))
+                    .skip(1)
+                    .take(self.block_size)
+                    .map(|v| format!("{v}"))
+                    .join(", ");
+                let column_str = format!(
+                    "... {} | {} | {} ...",
+                    values[0],
+                    block_list,
+                    values[self.block_size + 1]
+                );
+                log::debug!("Column {column_name} is not stackable:\n{column_str}");
+            }
+
+            stackable
+        });
+
+        match can_stack {
+            true => Ok(()),
+            false => Err("Block machine shape does not allow stacking".to_string()),
+        }
+    }
+
+    /// If any machine call could not be completed, that's bad because machine calls typically have side effects.
+    /// So, the underlying lookup / permutation / bus argument likely does not hold.
+    /// This function checks that all machine calls are complete, at least for a window of <block_size> rows.
+    fn check_incomplete_machine_calls(&self, complete: &HashSet<(u64, i32)>) -> Result<(), String> {
+        let machine_calls = self
+            .machine_parts
+            .identities
+            .iter()
+            .filter(|id| is_machine_call(id));
+
+        let incomplete_machine_calls = machine_calls
+            .flat_map(|call| {
+                let complete_rows = self
+                    .row_range()
+                    .filter(|row| complete.contains(&(call.id(), *row)))
+                    .collect::<Vec<_>>();
+                // Because we process rows -1..block_size+1, it is fine to have two incomplete machine calls,
+                // as long as <block_size> consecutive rows are complete.
+                if complete_rows.len() >= self.block_size {
+                    let is_consecutive = complete_rows.iter().max().unwrap()
+                        - complete_rows.iter().min().unwrap()
+                        == complete_rows.len() as i32 - 1;
+                    if is_consecutive {
+                        return vec![];
+                    }
+                }
+                self.row_range()
+                    .filter(|row| !complete.contains(&(call.id(), *row)))
+                    .map(|row| (call, row))
+                    .collect::<Vec<_>>()
             })
-            .any(|(identity, row)| !complete.contains(&(identity.id(), row)));
+            .collect::<Vec<_>>();
 
-        match has_incomplete_machine_calls {
-            true => Err("Incomplete machine calls".to_string()),
-            false => Ok(()),
+        if !incomplete_machine_calls.is_empty() {
+            Err(format!(
+                "Incomplete machine calls:\n  {}",
+                incomplete_machine_calls
+                    .iter()
+                    .map(|(id, row)| format!("{id} (row {row})"))
```

Comment:
Do we have access to the actual identity and should we print it here (instead of its id)?

Reply:
That's what this code does. `id` is for "identity" ;) I changed the variable name to make this clear.

Reply:
haha ok :)